### PR TITLE
added stoichiometric matrix and associated properties to Composite

### PR DESCRIPTION
**Pull request by bobmyhill**
This PR adds Composite properties which correspond to the stoichiometric matrix and associated column, row and nullspaces. These are useful for equilibrium calculations. 

This PR requires use of the pivots from Gaussian elimination. The necessary functions from sympy have been changing very quickly recently. For this PR, those functions are copied into reductions.py. We should be able to remove this code when sympy settles down. 

Secondly, the calculations are relatively expensive and rely on some optional parameters being set. For this reason, we only want to calculate the matrices once when they are first required, and then store the properties forever (i.e. we don't want to have to recalculate them every time we set_state or set_fractions). For this reason, I've also added a cached_property decorator. This can be replaced by the functools.cached_property decorator when we require Python 3.8+.

### CODE REVIEW

File: burnman/composite.py [hunk 1]
```diff
@@ -1,35 +1,37 @@
...
-        if len(fractions) < 1:
-            raise Exception('ERROR: we need at least one phase')
-
-        if len(phases) != len(fractions):
-            raise Exception(
-                'ERROR: different array lengths for phases and fractions')
 
-        total = sum(fractions)
-        if abs(total - 1.0) > 1e-10:
+def check_pairs(phases, fractions):
```
Comment (bobmyhill) [inline burnman/composite.py:19]: this change is just fixing indentation

**Comment by sannecottaar** [comment]
Looks good to me. I assume you have examples to commit to illustrate the use? The tests are slightly illustrative :)

**Comment by bobmyhill** [comment]
I do have some examples in mind. Do you want them in this PR or the next?

**Comment by sannecottaar** [comment]
Next is obviously fine.

**Pull request merged**